Update contributing guide (#1563)
iTwin/stratakit: STYLEGUIDE.md

# Style Guide

This style guide describes conventions and best practices for the StrataKit project.

> ## TSX

Reviewer: This section could mention the use of our custom `forwardRef` utility over `React.forwardRef`.

Author: Updated in https://github.com/iTwin/stratakit/pull/1563/commits/b7ae630a5435fe3c713109f8e20f96c2de4ab706

Reviewer: Correct commit is [`170fa17`](https://github.com/iTwin/stratakit/pull/1563/commits/170fa17f995c6d46d21a80db3a74628501fe3fb0).

### Imports

> - For Material UI imports use `import Rating from "@mui/material/Rating"` instead of `import { Rating } from "@mui/material"`. For more information see [MUI's guide on reducing bundle size](https://mui.com/material-ui/guides/minimizing-bundle-size/#avoid-barrel-imports)

Reviewer: I think this could be handled by the linter instead of through docs?

That's what we do for Ariakit imports.

Author: Linter is a good idea. Probalby best to do both since the linter doesn't provide any explaination about why we made that decision.

Author: Added linting in https://github.com/iTwin/stratakit/pull/1588

Reviewer: > since the linter doesn't provide any explaination about why we made that decision.

It could provide an explanation though. We can include any message we want.

Author: Removed imports section in https://github.com/iTwin/stratakit/pull/1588 and put comment in linter message in https://github.com/iTwin/stratakit/pull/1588/changes/18bb24cb922840530f5bfaa0221316e3cd1ed541

- For React imports use `import * as React from "react"`.
- Keep imports organized via running `pnpm run lint --write`

## CSS

> ### Nest CSS rules

Reviewer: "Nest CSS rules" is too broad. Not everything should be nested.

Author: Updated in https://github.com/iTwin/stratakit/pull/1563/commits/b7ae630a5435fe3c713109f8e20f96c2de4ab706

Reviewer: Correct commit is [`170fa17`](https://github.com/iTwin/stratakit/pull/1563/commits/170fa17f995c6d46d21a80db3a74628501fe3fb0).

[Nest CSS rules](https://developer.mozilla.org/en-US/docs/Web/CSS/Guides/Nesting/Using) under the main subject (selector to identify the DOM element). If there are multiple DOM elements (eg. tab list and tabs), use multiple groups.

**Do**

```css
.MuiToggleButton-root {
	// ...

	&:where(.Mui-selected) {
		// ...
	}
}
```

**Don't**

```css
.MuiToggleBottom-root {
	// ...
}

.MuiTogglButton-root:where(.Mui-selected) {
	//...
}
```

### Lower specificity

> Use [:where](https://developer.mozilla.org/en-US/docs/Web/CSS/Reference/Selectors/:where) to reduce specificity of rules. This makes it easier for consumers who are not using layers to override the StrataKit styling if need.

Reviewer: This guidance is also too broad. We want to keep the specificity flat, rather than simply "lowering" it.

For example, we don't want zero specificity using `:where(.MuiRating-root.Mui-disabled)`.

Author: Updated in https://github.com/iTwin/stratakit/pull/1563/commits/b7ae630a5435fe3c713109f8e20f96c2de4ab706

Reviewer: Correct commit is [`170fa17`](https://github.com/iTwin/stratakit/pull/1563/commits/170fa17f995c6d46d21a80db3a74628501fe3fb0).

| Selector                               | Specificity |          |
| -------------------------------------- | ----------- | -------- |
| `.MuiRating-root.Mui-disabled`         | `0, 2, 0`   | ❌ Don't |

> | `.MuiRating-root:where(.Mui-disabled)` | `0, 1, 0`   | ✅ Do    |

Reviewer: This doesn't need to be a table. It can be simple code snippets for consistency with other sections.

Author: Updated in https://github.com/iTwin/stratakit/pull/1563/commits/b7ae630a5435fe3c713109f8e20f96c2de4ab706

Reviewer: Correct commit is [`170fa17`](https://github.com/iTwin/stratakit/pull/1563/commits/170fa17f995c6d46d21a80db3a74628501fe3fb0).

### Use CSS logical properties

Use [CSS logical properties](https://developer.mozilla.org/en-US/docs/Web/CSS/Guides/Logical_properties_and_values) instead of physical properties. This is to support different locales that do use a left-right top-bottom layout.

Below are some examples. Refer to the documentation logical properties for a complete list.

| Physical property | Logical property    |
| ----------------- | ------------------- |
| margin-top        | margin-block-start  |
| margin-bottom     | margin-block-end    |
| margin-left       | margin-inline-start |
| magin-right       | margin-inline-end   |
| max-width         | max-inline-size     |
| max-height        | max-block-size      |

### Wrap hover rules with media query

> Wrap any rules relating to hover with a media query. This prevents a bug on iOS where the hover state may get stuck on after a tap.

Reviewer: This is not an iOS bug. It's the expected behavior on touch devices. Also should mention which media query specifically.

```suggestion
Wrap any rules relating to hover with the [`any-hover` media query](https://developer.mozilla.org/en-US/docs/Web/CSS/Reference/At-rules/@media/any-hover). This prevents an undesirable situation on touch devices where the hover state may get stuck on after a tap.
```

Author: Updated in https://github.com/iTwin/stratakit/pull/1563/commits/b7ae630a5435fe3c713109f8e20f96c2de4ab706

Reviewer 2: I may have explained this as an iOS bug, my mistake.

btw I am not seeing this change yet.

Author: Hmm tought I had applied it, but re-applied in https://github.com/iTwin/stratakit/pull/1563/commits/2723933e143a82fbeae6e5610593145a6281a905

```css
@media (any-hover: hover) {

> 	:where(.MuiTableSortLabel-root:hover) & {

Reviewer: Simpler example

```suggestion
	&:where(:hover) {
```

Author: Updated in https://github.com/iTwin/stratakit/pull/1563/commits/b7ae630a5435fe3c713109f8e20f96c2de4ab706

Reviewer: Correct commit is [`170fa17`](https://github.com/iTwin/stratakit/pull/1563/commits/170fa17f995c6d46d21a80db3a74628501fe3fb0).

Author: Also https://github.com/iTwin/stratakit/pull/1563/commits/81c326dac7c1b436012ae9fcab2a50532334cb20

# Style Guide

This style guide describes conventions and best practices for the StrataKit project.

> ## TSX

Reviewer: We use `interface` instead of `type`. Worth mentioning here.

You could borrow/adapt some language from https://ariakit.com/guide/coding-guidelines#prefer-interface-over-type

Author: Added in https://github.com/iTwin/stratakit/pull/1563/commits/8dda0dba5f5e6995b2e8239fc47bb2d7fdafff2b

# Style Guide

This style guide describes conventions and best practices for the StrataKit project.

## TSX

### forwardRef

Use the `forwardRef` utility from [packages/foundations](./packages/foundations/src/~utils.tsx) instead of `React.forwardRef`. The internal version allows refs to be loosely typed as `HTMLElement`.

## CSS

### Group CSS rules by target element

[Nest CSS rules](https://developer.mozilla.org/en-US/docs/Web/CSS/Guides/Nesting/Using) under the main subject (selector to identify the DOM element). If there are multiple DOM elements (eg. tab list and tabs), use multiple groups.

**✅ Do**

```css
.MuiToggleButton-root {
	// ...

	&:where(.Mui-selected) {
		// ...
	}
}
```

**❌ Don't**

```css
.MuiToggleBottom-root {
	// ...
}

> .MuiTogglButton-root:where(.Mui-selected) {

Reviewer 3: ```suggestion
.MuiToggleButton-root:where(.Mui-selected) {
```

	//...
}
```

### Target 0,1,0 sepecificity with rules

> Use [:where](https://developer.mozilla.org/en-US/docs/Web/CSS/Reference/Selectors/:where) to try and keep a flat specifcificy of 0,1,0. This makes it easier for consumers who are not using layers to override the StrataKit styling if need.

Reviewer 3: Typo and a link to mdn. I'm not sure if the last statement is correct: "This makes it easier for consumers who are not using layers to override the StrataKit styling if need." (lower specificity is mostly for our own benefit) 

```suggestion
Use [:where](https://developer.mozilla.org/en-US/docs/Web/CSS/Reference/Selectors/:where) to keep a flat [specificity](https://developer.mozilla.org/en-US/docs/Web/CSS/Guides/Cascade/Specificity) of 0,1,0. This makes it easier for consumers who are not using layers to override the StrataKit styling if need.
```

Author: > lower specificity is mostly for our own benefit

Can you elaborate on this?

Reviewer: Specificity only matters within a layer and does not impact consumers.

Consumer styles will be either unlayered, or in their own layer. Custom styles should never be inside `@layer stratakit`.

```suggestion
Use [:where](https://developer.mozilla.org/en-US/docs/Web/CSS/Reference/Selectors/:where) to keep a flat [specificity](https://developer.mozilla.org/en-US/docs/Web/CSS/Guides/Cascade/Specificity) of 0,1,0.
```

Author: If we aren't concerned about the consumer ability to override the style, what is the benefit to keeping it flat?

`.MuiRating-root.Mui-disabled` is easier to read than `.MuiRating-root:where(.Mui-disabled)` and how I would expect the rule to normally be written.

Reviewer: It's solely for authoring benefit. It's easier to manage a flat specificity, as you can read the CSS from top to bottom, knowing the declarations further down take priority.

Author: https://github.com/iTwin/stratakit/pull/1563/commits/0780437db3c8142021a561b9a52ba0f173fefc6d

You can use `:where`

> 

Reviewer 3: ```suggestion

```

Author: What was the intent of this comment?

Reviewer 3: The suggestion was to remove the line entirely, since this repeats the main sentence:
```
Use :where to try and keep <...>

You can use `:where`
```

Author: https://github.com/iTwin/stratakit/pull/1563/commits/0780437db3c8142021a561b9a52ba0f173fefc6d

**✅ Do**

```css
.MuiRating-root:where(.Mui-disabled) {
	/* 0,1,0 specificity */
}
```

**❌ Don't**

```css
.MuiRating-root.Mui-disabled {
	/* 0,2,0 specificity */
}
```

### Use CSS logical properties

Use [CSS logical properties](https://developer.mozilla.org/en-US/docs/Web/CSS/Guides/Logical_properties_and_values) instead of physical properties. This is to support different locales that do use a left-right top-bottom layout.

Wrap any rules relating to hover with the [`any-hover` media query](https://developer.mozilla.org/en-US/docs/Web/CSS/Reference/At-rules/@media/any-hover). This prevents an undesirable situation on touch devices where the hover state may get stuck on after a tap.

> 

Reviewer 3: duplicate

```suggestion

```

Author: Removed in https://github.com/iTwin/stratakit/pull/1563/commits/d65e4864a507b69badc2e320860dab69d1b895c5

| Physical property | Logical property    |
| ----------------- | ------------------- |
| margin-top        | margin-block-start  |
| margin-bottom     | margin-block-end    |
| margin-left       | margin-inline-start |
| magin-right       | margin-inline-end   |
| max-width         | max-inline-size     |
| max-height        | max-block-size      |

### Wrap hover rules with media query

Wrap any rules relating to hover with the [`any-hover` media query](https://developer.mozilla.org/en-US/docs/Web/CSS/Reference/At-rules/@media/any-hover). This prevents an undesirable situation on touch devices where the hover state may get stuck on after a tap.

```css
@media (any-hover: hover) {

> 	:where(:hover) & {

Reviewer 3: nit: more common use-case

```suggestion
	&:where(:hover) {
```

Reviewer: I made the exact same suggestion in https://github.com/iTwin/stratakit/pull/1563#discussion_r3475920921. Weird.

Author: Updated in https://github.com/iTwin/stratakit/pull/1563/commits/81c326dac7c1b436012ae9fcab2a50532334cb20

# Style Guide

This style guide describes conventions and best practices for the StrataKit project.

## TSX

### Use `interface` instead `type`

Although they are very similar, use `interface` for consistency and [better detection of type errors](https://ariakit.com/guide/coding-guidelines#prefer-interface-over-type).

### forwardRef

Use the `forwardRef` utility from [packages/foundations](./packages/foundations/src/~utils.tsx) instead of `React.forwardRef`. The internal version allows refs to be loosely typed as `HTMLElement`.

## CSS

### Group CSS rules by target element

> [Nest CSS rules](https://developer.mozilla.org/en-US/docs/Web/CSS/Guides/Nesting/Using) under the main subject (selector to identify the DOM element). If there are multiple DOM elements (eg. tab list and tabs), use multiple groups.

Reviewer: ```suggestion
[Nest CSS rules](https://developer.mozilla.org/en-US/docs/Web/CSS/Guides/Nesting/Using) under the main subject (i.e. the selector to identify the DOM element). If there are multiple DOM elements (e.g. tab list and tabs), use multiple groups.
```

Author: Updated in https://github.com/iTwin/stratakit/pull/1563/commits/4dd28e91426ef18694ba250ed4887903106d2775

**✅ Do**

```css
.MuiToggleButton-root {
	// ...

	&:where(.Mui-selected) {
		// ...
	}

> }

Reviewer: It would be good to show an example of multiple blocks targeting different elements.

```suggestion
.MuiButton-root {
	// ...

	&:where(.Mui-disabled) {
		// ...
	}
}

.MuiButton-icon {
	// ...
}
```

Author: Updated in https://github.com/iTwin/stratakit/pull/1563/commits/d16094c5a0aa2e8470ba63ab0e1f1ec0b207a33e

```

**❌ Don't**

```css
.MuiToggleBottom-root {
	// ...
}

.MuiToggleButton-root:where(.Mui-selected) {
	//...

> }

Reviewer: It would be good to show an example of nested element selectors to avoid.

```suggestion
.MuiButton-root {
	// ...

	.MuiButton-icon {
		// ...
	}
}

.MuiButton-root:where(.Mui-disabled) {
	// ...
}
```

Author: Updated in https://github.com/iTwin/stratakit/pull/1563/commits/d16094c5a0aa2e8470ba63ab0e1f1ec0b207a33e

```

### Target 0,1,0 sepecificity with rules

Use [:where](https://developer.mozilla.org/en-US/docs/Web/CSS/Reference/Selectors/:where) to try and keep a flat specifcificy of 0,1,0. This makes it easier for consumers who are not using layers to override the StrataKit styling if need.

You can use `:where` to target selectors without increasing the specificity.

**✅ Do**

```css
.MuiRating-root:where(.Mui-disabled) {
	/* 0,1,0 specificity */

> }

Reviewer: This should be nested.

```suggestion
.MuiRating-root {
	&:where(.Mui-disabled) {
		/* 0,1,0 specificity */
	}
}
```

Author: Updated in https://github.com/iTwin/stratakit/pull/1563/commits/b21d694d69e32b2b9b63b14356ca04693463b518

# Style Guide

This style guide describes conventions and best practices for the StrataKit project.

## TSX

### Imports

- For Material UI imports use `import Rating from "@mui/material/Rating"` instead of `import { Rating } from "@mui/material"`. For more information see [MUI's guide on reducing bundle size](https://mui.com/material-ui/guides/minimizing-bundle-size/#avoid-barrel-imports)
- For React imports use `import * as React from "react"`.
- Keep imports organized via running `pnpm run lint --write`

### forwardRef

> Use the `forwardRef` from [packages/foundations](./packages/foundations/src/~utils.tsx) instead of `React.forward`. The internal version allows refs to be loosely typed as `HTMLElement`.

Reviewer: It's forward**Ref**

```suggestion
Use the `forwardRef` utility from [packages/foundations](./packages/foundations/src/~utils.tsx) instead of `React.forwardRef`. The internal version allows refs to be loosely typed as `HTMLElement`.
```

> Use the `forwardRef` from [packages/foundations](./packages/foundations/src/~utils.tsx) instead of `React.forwarrd`. The internal version allows refs to be loosely typed as `HTMLElement`.

Reviewer 2: ```suggestion
Use the `forwardRef` from [packages/foundations](./packages/foundations/src/~utils.tsx) instead of `React.forward`. The internal version allows refs to be loosely typed as `HTMLElement`.
```
